README.md
CratesPlus
=====================

CratesPlus is a highly used and feature rich crates plugin, that supports Bukkit + Spigot from version 1.7 all the way to 1.14.2.

## Development Build
[![CircleCI](https://circleci.com/gh/ConnorLinfoot/CratesPlus/tree/master.svg?style=svg)](https://api.linfoot.dev/v2/ci/artifact/CratesPlus/latest/download)

The latest dev build of this can be downloaded from the link on the image above. Please note this version can be buggy and will have uncompleted features and some features may be removed before release. Please backup your config before using a dev build!

### CratesPlus 5.0 Important Notice

The latest development builds are for CratesPlus 5.0, I can NOT stress enough that this version has some major changes which are likely to break things during development. Please don't use this on a live/public server.

## Stable Builds
You can download stable builds from the [Spigot versions page](https://www.spigotmc.org/resources/cratesplus-custom-crates-free.5018/history).

## Wiki
For full information, commands and features visit the [Wiki](https://github.com/ConnorLinfoot/CratesPlus/wiki).

## Issues
Please report any issues into the [GitHub issues](https://github.com/ConnorLinfoot/CratesPlus/issues). Do not post them on the Spigot thread or PM me.
You can use '/crate debug' to generate a copy of all the data I should need to assist with the issue.

If you report an issue which has been fixed or does not contain the information required it may be closed without any notification.


## Changelog
To see the full changelog for all publicly released versions checkout the [Changelog File](https://github.com/ConnorLinfoot/CratesPlus/blob/master/Changelog.md).


Changelog.md
**4.5.3** *(Nov 15 2020)* - [Download](https://www.spigotmc.org/resources/5018/download?version=367260)
- Feature: Support for Bukkit/Spigot 1.16.4. Other versions should work from 1.8 to 1.16, but only 1.8 and 1.16.4 were specifically tested.
- Bug: Fixed an issue in some versions of Spigot where interacting with chests would throw an exception.
- Bug: Fixed the sign UI not working in most cases.
- Note: This version now requires Java 8, although this should impact next to nobody.

**4.5.2** *(Jun 1 2019)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-13-1.5018/download?version=276471)
- Feature: Support for Bukkit/Spigot 1.14 - 1.14.2.

**4.5.1** *(Sep 17 2018)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-13-1.5018/download?version=240597)
- Bug: Fixed an issue that caused opening crates to not work on some 1.13.1 builds. This was due to the display name no longer returning its color codes, I am not yet sure if this is intended but this build does fix it in the meantime.

**4.5** *(Sep 16 2018)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-12-2.5018/download?version=240554)
- Feature: Support for Bukkit/Spigot 1.13.
- Removal: Removed support for winning `Item ID`, use `Item Type` instead.
- Removal: Removed support for Enchantment ID's, use Enchantment names instead.

**4.4.5** *(Feb 14 2018)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-12-2.5018/download?version=208286)
- Bug: Updated the URL for Spigot update checks as the old one is now disabled.

**4.4.4** *(Sep 25 2017)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-12-2.5018/download?version=186440)
- Feature: Support for Bukkit/Spigot 1.12.2.

**4.4.3** *(Aug 3 2017)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-12-1.5018/download?version=176069)
- Feature: Support for Bukkit/Spigot 1.12.1.

**4.4.2** *(Jul 18 2017)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-12.5018/download?version=172748)
- Bug: Fixed an error if you had more than 54 items to preview. This will include pagination in a future update.
- Bug: Added a null check to prevent an NPE if you didn't define winnings.
- Bug: Now forcing item types to be uppercase for commands just like items.

**4.4.1** *(Jun 9 2017)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-12.5018/download?version=164332)
- Feature: Support for Bukkit/Spigot 1.12.

**4.4** *(Apr 17 2017)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-11-2.5018/download?version=155087)
- Feature: You can now set the lore of a key in the config file. (Warning: changing the lore WILL break existing keys given to players)
- Feature: Added support for storing all meta data when using the built in GUI for managing winnings. This includes enchantments, potions, books, player heads etc.
- Feature: Added support for ItemFlags (1.9+). Check the list of [Flags](https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/inventory/ItemFlag.html).
- Feature: Added new command placeholder "rand" for generating a random number (Example: %rand;100;500% would generate a number between 100 and 500)
- Improvement: Added %displayname% placeholder to commands.
- Improvement: Changed "BasicGUI" default length to 5.
- Bug: Fixed NPE with player interact.
- Bug: Fixed spawn eggs not working in 1.11+

**4.3.3** *(Apr 9 2017)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-11-2.5018/download?version=153094)
- Bug: Removed left over debug from last update.

**4.3.2** *(Apr 9 2017)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-11-2.5018/download?version=153092)
- Bug: Fixed an issue that allowed people to use keys multiple times.

**4.3.1** *(Apr 6 2017)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-11-2.5018/download?version=152479)
- Bug: Fixed issue where keys sometimes didn't have a lore set. Causing said keys to not work.

**4.3** *(Apr 1 2017)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-11-2.5018/download?version=151600)
- Improvement: Now supports "Block Data" for setting the crates data ID.
- Improvement: Debug link now uses v2 of MC Debug, will allow the link that is shared to be viewed easier and even managed.
- Improvement: Keys are now checked with their lore as well as the name, this should help prevent plugins with things like /rename exploiting this.
- Improvement: Added a custom metrics class, will be used with a new MC Stats tool coming soon.
- Improvement: Keys lost on death will now be put in the claim GUI instead.
- Improvement: Better handling of rename for 1.7.
- Improvement: I really like purple.
- Improvement: No but seriously, I ***really*** like purple. 
- Bug: Fixed issue with debug link and config backups, now using GitHub Gist instead of Hastebin.
- Bug: Hologram warning now only shows up once if it finds an incompatible version.

**4.2.4** *(Dec 21 2016)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-11-1.5018/download?version=131086)
- Feature: Support for Bukkit/Spigot 1.11.2.

**4.2.3** *(Dec 20 2016)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-11-1.5018/download?version=130903)
- Feature: Support for Bukkit/Spigot 1.11.1.

**4.2.2** *(Dec 11 2016)* - [Download](https://www.spigotmc.org/resources/cratesplus-free-crates-plugin-1-7-1-10-2.5018/download?version=129396)
- Feature: Support for Bukkit/Spigot 1.11.
- Bug: Fixed command being ran async.
- Bug: Fixed issue with location being null stopping plugin starting correctly.
- Bug: Fixed block break on 1.7 (hopefully, not actually tested this)


src/main/java/plus/crates/Utils/GUI.java
package plus.crates.Utils;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.*;

/**
 * Connor Linfoot's GUI Library
 *
 * @version 4.0
 */
public class GUI {
    public static HashMap<UUID, GUI> guis = new HashMap<>();
    public static HashMap<UUID, Integer> pageTracker = new HashMap<>();
    public static ArrayList<UUID> ignoreClosing = new ArrayList<>();
    private static final HashMap<Integer, Integer> itemMapping = new HashMap<Integer, Integer>() {{
        put(0, 10);
        put(1, 11);
        put(2, 12);
        put(3, 13);
        put(4, 14);
        put(5, 15);
        put(6, 16);
        put(7, 19);
        put(8, 20);
        put(9, 21);
        put(10, 22);
        put(11, 23);
        put(12, 24);
        put(13, 25);
        put(14, 28);
        put(15, 29);
        put(16, 30);
        put(17, 31);
        put(18, 32);
        put(19, 33);
        put(20, 34);
    }};

    private String title;
    private NavigableMap<Integer, ItemStack> items = new TreeMap<>();
    private HashMap<Integer, ClickHandler> clickHandlers = new HashMap<>();
    private ClickHandler goBackHandler = null;
    private boolean showPages = false;

    public GUI(String title) {
        this.title = title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getTitle() {
        return title;
    }

    public NavigableMap<Integer, ItemStack> getItems() {
        return items;
    }

    public void addItem(ItemStack itemStack) {
        addItem(itemStack, null);
    }

    public void addItem(ItemStack itemStack, ClickHandler clickHandler) {
        setItem(getItems().isEmpty() ? 0 : getItems().lastKey() + 1, itemStack, clickHandler);
    }

    public void setItem(Integer slot, ItemStack itemStack) {
        setItem(slot, itemStack, null);
    }

    public void setItem(Integer slot, ItemStack itemStack, ClickHandler clickHandler) {
        items.put(slot, itemStack);
        clickHandlers.put(slot, clickHandler);
    }

    private Integer getPlayerPage(Player player) {
        if (pageTracker.containsKey(player.getUniqueId()))
            return pageTracker.get(player.getUniqueId());
        return 1;
    }

    public void handleClick(Player player, Integer slot) {
        if (slot == 48 && getPlayerPage(player) <= 1 && getGoBackHandler() != null) {
            getGoBackHandler().doClick(player, this);
        } else if (slot == 48 && getPlayerPage(player) > 1) {
            GUI.ignoreClosing.add(player.getUniqueId());
            open(player, getPlayerPage(player) - 1);
        } else if (slot == 50) {
            GUI.ignoreClosing.add(player.getUniqueId());
            open(player, getPlayerPage(player) + 1);
        } else {
            Integer item = calculateItemFromSlot(player, slot);
            if (clickHandlers.containsKey(item) && clickHandlers.get(item) != null) {
                ClickHandler clickHandler = clickHandlers.get(item);
                clickHandler.doClick(player, this);
            }
        }
    }

    private Integer calculateItemFromSlot(Player player, Integer slot) {
        int actualI = 0;
        Integer page = getPlayerPage(player);

        if (page > 1) {
            actualI = 21 * (page - 1);
        }

        // We start at slot 10
        for (int i = 10; i < 54; i++) {
            if (i == 17 || i == 18 || i == 26 || i == 27 || i >= 35)
                continue;
            if (i == slot) {
                // GOT IT
                return actualI;
            }
            actualI++;
        }
        return null;
    }

    public boolean isShowPages() {
        return showPages;
    }

    public void setShowPages(boolean showPages) {
        this.showPages = showPages;
    }

    public ClickHandler getGoBackHandler() {
        return goBackHandler;
    }

    public void setGoBackHandler(ClickHandler goBackHandler) {
        this.goBackHandler = goBackHandler;
    }

    protected Inventory create(Integer page) {
        if (page < 1) page = 1;
        int itemsPerPage = 21;
        int pages = 1;
        if (!getItems().isEmpty())
            pages = (int) Math.ceil(getItems().lastKey() / itemsPerPage) + 1;
        int startFrom = ((page - 1) * itemsPerPage);

        int size = 45;
        if (getGoBackHandler() != null || page > 1 || pages > page)
            size = size + 9;

        String title = getTitle();
        if (isShowPages()) {
            title += " (Page " + page + "/" + pages + ")";
        }
        Inventory inventory = Bukkit.createInventory(null, size, title);

        for (int i = startFrom; i < startFrom + itemsPerPage; i++) {
            if (!getItems().containsKey(i))
                continue;
            Integer slot = itemMapping.get(i - ((page - 1) * itemsPerPage));
            inventory.setItem(slot, getItems().get(i));
        }

        if (page == 1 && getGoBackHandler() != null) {
            ItemStack back = new ItemStack(Material.ARROW);
            ItemMeta backMeta = back.getItemMeta();
            backMeta.setDisplayName(ChatColor.YELLOW + "Go Back");
            back.setItemMeta(backMeta);
            inventory.setItem(48, back);
        } else if (page > 1) {
            ItemStack prev = new ItemStack(Material.ARROW);
            ItemMeta prevMeta = prev.getItemMeta();
            prevMeta.setDisplayName(ChatColor.YELLOW + "Previous Page");
            prev.setItemMeta(prevMeta);
            inventory.setItem(48, prev);
        }

        if (pages > page) {
            ItemStack prev = new ItemStack(Material.ARROW);
            ItemMeta prevMeta = prev.getItemMeta();
            prevMeta.setDisplayName(ChatColor.YELLOW + "Next Page");
            prev.setItemMeta(prevMeta);
            inventory.setItem(50, prev);
        }

        return inventory;
    }

    public void open(Player player) {
        open(player, 1);
    }

    public void open(Player player, Integer page) {
        guis.put(player.getUniqueId(), this);
        pageTracker.put(player.getUniqueId(), page);
        player.openInventory(create(page));
    }

    public abstract static class ClickHandler {
        public abstract void doClick(Player player, GUI gui);
    }

}